### PR TITLE
add `ENABLED` and `\cdot` front-end

### PR DESCRIPTION
**Pull request by johnyf**
These changes:

- add the following proof directives to the module `library/TLAPS.tla`:
  - `ENABLEDaxioms`: apply proof rules about the operator `ENABLED` (implemented in the module `src/expr/e_action.ml`), and ensure for soundness that assumptions are of appropriate expression level (implemented in the module `src/module/m_elab.ml`)
  - `ExpandENABLED`: replace the operator `ENABLED` with rigid quantification of primed occurrences of variables (implemented in the module `src/expr/e_action.ml`)
  - `ExpandCdot`: replace the operator `\cdot` with rigid quantification over primed and unprimed occurrences of variables (implemented in the module `src/expr/e_action.ml`)
  - `AutoUSE`: expand defined operators where needed to soundly replace the operators `ENABLED` and `\cdot` with rigid quantification, raise errors for declared operators and parameters of operators (implemented in the module `src/expr/e_action.ml`)
  - `Lambdify`: replace the operator `ENABLED` and `\cdot` with constant-level expressions that include `LAMBDA` expressions with normalized parameter names (implemented in the module `src/expr/e_action.ml`)
  - `LevelComparison`: prove sequents with modified levels of declared operators, for example replace an `ACTION` declaration with a `STATE` declaration (implemented in the module `src/expr/e_level_comparison.ml`)
- computation of TLA+ expression levels (implemented in the module `src/expr/e_levels.ml`)
- add version 13 of the fingerprints file format, which includes the newly implemented proof directives (implemented in the module `src/backend/fpfile.ml`)
- add command-line option `--line <int>` for line to prove (abbreviates `--toolbox n n`) (implemented in the module `src/tlapm_args.ml`)
- add command-line option `--smt-logic <logic>` for setting the SMT solver logic, with default the logic UFNIA (implemented in the modules `src/backend/smt.ml`, `src/params.ml`, `src/tlapm_args.ml`)
- save fingerprints and auxiliary files that correspond to the file `filename.tla` in the directory `__tlacache__/filename.tlaps` (implemented in the module `src/tlapm.ml`, updated test file `test/fast/fingerprint/load_v8old_test.tla`)
- save kind of declaration (`CONSTANT`, `STATE`, `ACTION`, `TEMPORAL`, unknown) in fingerprints (implemented in the module `src/backend/fingerprints.ml`)
- save proof directives that appear in `BY` statements in the fingerprints, to ensure that proof obligations with the same assertions and different `BY` statements result in different fingerprints (implemented in the module `src/backend/fingerprints.ml`)
- add tests for the newly implemented proof directives to the directory `test/fast/enabled_cdot/`:
  - `Cdot_test.tla`
  - `ENABLEDaxioms_test.tla`
  - `Enabled_test.tla`
  - `ExpandENABLED_LET_test.tla`
  - `ExpandOnlyCdot_test.tla`
  - `ExpandOnlyENABLED_test.tla`
  - `Level_test.tla`
  - `NestedCdot_test.tla`
  - `NestedENABLED_from_AutoUSE_test.tla`
  - `NestedENABLED_test.tla`
- bug fixes to coalescing (in the module `src/frontend/coalesce.ml`)
- prepend `CONSTANT_`, `STATE_`, `ACTION_`, `TEMPORAL_`, `Unknown_`, `VARIABLE_` to coalesced identifier, to represent level information about declared and defined operators and avoid related unsoundness bugs (implemented in `src/frontend/coalesce.ml`)
- reimplement substitutivity computation (in the module `src/expr/e_substitutive.ml`)
- add class `map` to module `src/expr/e_subst.ml` (to enable subclassing)
- add class `map_visible_hyp` to module `src/expr/e_subst.ml` (substitution that visits only visible hypotheses in a sequent's context)
- add classes `map_visible_hyp`, `iter_visible_hyp`, `map_rename` to the module `src/expr/e_visit.ml`
- add functions `set_to_list`, `collect_unprimed_vars`, `collect_primed_vars` to the module `src/expr/e_visit.ml`
- call function `map_visible_hyp` from module `src/frontend/action.ml`
- call level computation in `src/frontend/action.ml`, instead of `Expr.Constness.is_const`
- bug fix in `src/frontend/action.ml` to add `ENABLED` in pattern case
- bug fix in `src/frontend/action.ml` to assert false for the operators `UNCHANGED`, `\cdot`, `~>`, `-+->`, `[]`, `<>`, `\AA`, `\EE`, subscript, temporal subscript, `WF_`, `SF_` in pattern case
- add comments to the types in the module `src/expr/e_t.ml`
- add functions `visibility_to_string`, `print_cx`, `cx_front`, `scx_front`, `find_hyp_named`, `expr_locus`, `format_locus`, `shape_to_arity`, `number_of_hyp`, `sequent_stats` to the module `src/expr/e_t.ml`
- add the example of incrementing an integer-valued variable `examples-draft/SimpleExampleWF.tla`
- activate proving for submodules (implemented in the module `src/module/m_parser.ml`)
- modules from `INSTANCE` and `EXTENDS` statements could be submodules of a module that the module extends, and is itself another file (implemented in the module `src/module/m_dep.ml`, `src/module/m_save.ml`)
- refactor module `src/module/m_elab.ml`, part moved to the module `src/module/m_subst.ml` (described below)
- add visitor of TLA+ module syntax graphs (implemented in the module `src/module/m_visit.ml`)
- add utilities for performing substitutions in TLA+ module syntax graphs (implemented in the module `src/module/m_subst.ml`)
- refactor `src/tlapm.ml`, change error messages, printing of count of failed obligations
- `failwith_unsupp` for the operators `~>`, `ENABLED`, `\cdot`, `-+->`, `[]`, `<>` in the Isabelle and Zenon backends (implemented in the modules `src/backend/isabelle.ml`, src/backend/zenon.ml)
- coalesce defined operator occurrences, operator application, and first-order quantification in `src/backend/ls4.ml`
- update the modules `src/backend/prep.ml`, `src/method.ml`, `src/method_prs.ml` with the newly implemented proof directives
- add `~>` to temporal operators in pattern case in the module `src/expr/e_elab.ml`
- add function `first_n` to module `src/util/deque.ml`
- augment error message in the module `src/expr/e_deref.ml`
- `BSeq` was a binary operator (implemented in the module `src/module/m_standard.ml`)
- update the documentation file `todo.txt`
- update the file `src/Makefile`
- update the generated file `src/.depend`

### CODE REVIEW

**Comment by kape1395** [comment]
It fails to compile for me with error:
```
File "e_visit.ml", line 470, characters 14-33:
Error: Unbound value Hashtbl.to_seq_keys
```

**Comment by kape1395** [comment]
It works with ocaml 4.09.1 and fixes #8, at least for my case.

**Review by damiendoligez — COMMENTED** [review]
Looks good in general. I have a commented a few points of detail.

**Comment by johnyf** [comment]
Thank you for reviewing, I will address the comments.

File: src/expr/e_levels.ml
Comment (johnyf) [reply src/expr/e_levels.ml]: Thank you for the suggestion, I added it to the changes.

File: src/expr/e_levels.ml
Comment (johnyf) [reply src/expr/e_levels.ml:747]: Thank you for the suggestion, I added it to the changes.

File: src/expr/e_visit.ml
Comment (johnyf) [reply src/expr/e_visit.ml]: Thank you for the suggestion, I added it to the changes.

File: src/expr/e_visit.ml
Comment (johnyf) [reply src/expr/e_visit.ml]: Thank you for the suggestion, I added it to the changes.

File: src/expr/e_visit.ml
Comment (johnyf) [reply src/expr/e_visit.ml]: Thank you for the suggestion, I added it to the changes.

File: src/frontend/coalesce.ml
Comment (johnyf) [reply src/frontend/coalesce.ml]: Thank you for the suggestion, I added it to the changes.

File: src/frontend/coalesce.ml
Comment (johnyf) [reply src/frontend/coalesce.ml]: Thank you for the suggestion, I added it to the changes. (It appears that `args` as the second argument to the function `List.mapi` is needed.)

File: src/system.ml
Comment (johnyf) [reply src/system.ml]: Thank you for the suggestion, I added it to the changes.

File: src/system.ml
Comment (johnyf) [reply src/system.ml]: Thank you for the suggestion, I added it to the changes.

File: src/util/deque.ml
Comment (johnyf) [reply src/util/deque.ml]: Thank you for the suggestion, I added it to the changes.

File: src/util/deque.ml
Comment (johnyf) [reply src/util/deque.ml]: Thank you for the suggestion, I added it to the changes.

File: src/util/ext.ml
Comment (johnyf) [reply src/util/ext.ml]: Thank you for the suggestion, I added it to the changes.

**Review by johnyf — COMMENTED** [review]
I have added a few more commits to address the comments.

**Review by damiendoligez — APPROVED** [review]
This all looks good now. Thanks for the changes and explanations.

**Comment by damiendoligez** [comment]
This is good to merge now. @johnyf I'll let you push the merge button after making any last-minute changes, if you have any.

**Comment by johnyf** [comment]
Thank you for approving these changes, I merged them.

**Pull request merged**
